core/src/main/scala/org/apache/spark/deploy/history/FsHistoryProvider.scala

```diff
@@ -449,7 +450,7 @@ private[history] class FsHistoryProvider(conf: SparkConf, clock: Clock)
               listing.write(info.copy(lastProcessed = newLastScanTime, fileSize = entry.getLen()))
             }
 
-            if (info.fileSize < entry.getLen()) {
+            if (info.fileSize < entry.getLen() || checkAbsoluteLength(info, entry)) {
```

Comment:
I think we can create a function to get the length of given file:
1. If the new conf is enabled and the input is DFSInputStream, use `getFileLength` (or `max(getFileLength, entry.getLen()`)
2. otherwise `entry.getLen()`

The logic can be simpler.

Reply:
Have you looked @ this getFileLength() call to see how well it updates?
 
FwIW [HADOOP-15606](https://issues.apache.org/jira/browse/HADOOP-15606) proposes adding a method like this for all streams, though that proposal includes the need for specification and tests. Generally the HDFS team are a bit lax about that spec -> test workflow, which doesn't help downstream code or other implementations.

Reply:
Thanks @gengliangwang for looking into this. Here it doesn't need to check the checkAbsoluteLength if the FileStatus.getLen() is more than the cached fileSize, if we update to `max(getFileLength, entry.getLen()))` it checks the absoluteLength always which may not be necessary.

Reply:
Thanks @steveloughran for looking into this.
> Have you looked @ this getFileLength() call to see how well it updates?

I looked at the DFSInputStream.getFileLength() api, it gives locatedBlocks.getFileLength() + lastBlockBeingWrittenLength, here locatedBlocks.getFileLength() is the value got from NameNode for all the completed blocks and lastBlockBeingWrittenLength is the lastblock lenth from DataNode which is not the completed block.

> FwIW HADOOP-15606 proposes adding a method like this for all streams

Thanks for the pointer, once this is available we can update to use it.

Reply:
...there's no timetable for that getLength thing, but if HDFS already supports the API, I'm more motivated to implement it. It has benefits in cloud stores in general
1. saves apps going an up front HEAD/getFileStatus() to know how long their data is; the GET should return it.
2. for S3 Select, you get back the filtered data so don't know how much you will see until the GET is issued